fineract-provider/src/main/java/org/apache/fineract/cob/loan/FetchAndLockLoanTasklet.java

```diff
@@ -46,67 +40,44 @@ public class FetchAndLockLoanTasklet implements Tasklet {
-        List<Long> allNonClosedLoanIds = retrieveLoanIdService.retrieveLoanIdsNDaysBehind(NUMBER_OF_DAYS_BEHIND, businessDate);
-        if (allNonClosedLoanIds.isEmpty()) {
-            contribution.getStepExecution().getJobExecution().getExecutionContext().put(LoanCOBConstant.LOAN_IDS, Collections.emptyList());
+        LoanCOBMinANdMaxLoanId minAndMaxLoanId = retrieveLoanIdService.retrieveMinAndMaxLoanIdsNDaysBehind(NUMBER_OF_DAYS_BEHIND,
+                businessDate);
+        if (Objects.isNull(minAndMaxLoanId)
+                || (Objects.isNull(minAndMaxLoanId.getMinLoanId()) && Objects.isNull(minAndMaxLoanId.getMaxLoanId()))) {
+            contribution.getStepExecution().getJobExecution().getExecutionContext().put(LoanCOBConstant.LOAN_IDS, null);
             return RepeatStatus.FINISHED;
         }
-        List<Long> remainingIds = new ArrayList<>(allNonClosedLoanIds);
-
-        List<List<Long>> remainingIdPartitions = Lists.partition(remainingIds, getInClauseParameterSizeLimit());
-        List<LoanAccountLock> loanAccountLocks = new ArrayList<>();
-        remainingIdPartitions.forEach(
-                remainingIdPartition -> loanAccountLocks.addAll(loanAccountLockRepository.findAllByLoanIdIn(remainingIdPartition)));
-
-        List<Long> alreadySoftLockedAccounts = loanAccountLocks.stream()
-                .filter(e -> LockOwner.LOAN_COB_PARTITIONING.equals(e.getLockOwner())).map(LoanAccountLock::getLoanId).toList();
-        List<Long> alreadyMarkedForInlineCOBLockedAccounts = loanAccountLocks.stream()
-                .filter(e -> LockOwner.LOAN_INLINE_COB_PROCESSING.equals(e.getLockOwner())).map(LoanAccountLock::getLoanId).toList();
-        List<Long> alreadyMarkedForChunkProcessingLockedAccounts = loanAccountLocks.stream()
-                .filter(e -> LockOwner.LOAN_COB_CHUNK_PROCESSING.equals(e.getLockOwner())).map(LoanAccountLock::getLoanId).toList();
 
-        // Remove already hard locked accounts
-        remainingIds.removeAll(alreadyMarkedForChunkProcessingLockedAccounts);
-        remainingIds.removeAll(alreadyMarkedForInlineCOBLockedAccounts);
+        applySoftLock(businessDate);
 
-        List<Long> lockableLoanAccounts = new ArrayList<>(remainingIds);
-        lockableLoanAccounts.removeAll(alreadySoftLockedAccounts);
-
-        applySoftLock(lockableLoanAccounts);
-
-        contribution.getStepExecution().getJobExecution().getExecutionContext().put(LoanCOBConstant.LOAN_IDS, remainingIds);
+        contribution.getStepExecution().getJobExecution().getExecutionContext().put(LoanCOBConstant.LOAN_IDS, minAndMaxLoanId);
 
         return RepeatStatus.FINISHED;
     }
 
-    private void applySoftLock(List<Long> accountsToLock) {
+    private void applySoftLock(LocalDate businessDate) {
         LocalDate cobBusinessDate = ThreadLocalContextUtil.getBusinessDateByType(BusinessDateType.COB_DATE);
-        jdbcTemplate.batchUpdate("""
+        jdbcTemplate.update("""
```

Comment:
Aren't we missing the minLoanId and maxLoanId range filter in this query? In case there's a loan getting created between retrieving the min/max loanIds and the soft locking. The soft locking will take that loan too, but the min/max range will not include it, hence the COB will not pick it up.

Reply:
I added this filter